### PR TITLE
add Popup methods to manipulate container class names

### PR DESCRIPTION
**Pull request by Ashot-KR**
Added 3 methods to `Popup`:
`addClassName` – add className to container
`removeClassName` – remove className from container
`toggleClassName` – toggle className

All methods is just shortcut wrappers around `popup._container.classList` methods respectively.

Related to #8655

### CODE REVIEW

**Comment by Ashot-KR** [comment]
I agree.
I reverted initial `className` setup to `forEach` instead of `classList.add` with multiple arguments (yes, i forgot to check browser compatibility before i did that :bowtie:)
`addClassName`, `removeClassName` and `toggleClassName` now accepts only single `string` argument

**Comment by colleenmcginnis** [comment]
Thanks @mourner! Tagging in @danswick for docs review.

File: .gitignore [hunk 1]
```diff
@@ -25,3 +25,4 @@ _site
 yarn-error.log
 yarn-debug.log
 npm-debug.log
+.idea
```
Comment (vakila) [inline .gitignore]: Is this addition intentional/necessary?

File: src/ui/popup.js [hunk 2]
```diff
@@ -359,10 +400,10 @@ export default class Popup extends Evented {
             this._tip       = DOM.create('div', 'mapboxgl-popup-tip', this._container);
             this._container.appendChild(this._content);
             if (this.options.className) {
-                this.options.className.split(' ').forEach(name =>
-                    this._container.classList.add(name));
+                this.options.className.split(' ').forEach(name => {
+                    this._container.classList.add(name);
+                });
```
Comment (vakila) [inline src/ui/popup.js]: Is this change necessary, or perhaps just something left over from a previous change that modified this and was then reverted?

File: src/ui/popup.js [hunk 3]
```diff
@@ -332,6 +332,47 @@ export default class Popup extends Evented {
         return this;
     }
 
+    /**
+     * Adds class to container element.
```
Comment (vakila) [inline src/ui/popup.js]:
```suggestion
     * Adds a CSS class to the popup container element.
```

File: src/ui/popup.js [hunk 4]
```diff
@@ -332,6 +332,47 @@ export default class Popup extends Evented {
...
+     *
+     * @param {string} className non empty string
```
Comment (vakila) [inline src/ui/popup.js]:
```suggestion
     * @param {string} className Non-empty string with CSS class name to add to popup container 
```

File: src/ui/popup.js [hunk 5]
```diff
@@ -332,6 +332,47 @@ export default class Popup extends Evented {
...
+     * @example
+     * let popup = new mapboxgl.Popup()
+     * popup.addClassName('some-class')
+     */
+    addClassName(className: string) {
+        this._container.classList.add(className);
+    }
+
+    /**
+     * Removes class from container element.
```
Comment (vakila) [inline src/ui/popup.js]:
```suggestion
     * Removes a class from the popup container element.
```

File: src/ui/popup.js [hunk 6]
```diff
@@ -332,6 +332,47 @@ export default class Popup extends Evented {
...
+     *
+     * @param {string} className non empty string
```
Comment (vakila) [inline src/ui/popup.js]:
```suggestion
     * @param {string} className Non-empty string with CSS class name to remove from popup container 
```

File: src/ui/popup.js [hunk 7]
```diff
@@ -332,6 +332,47 @@ export default class Popup extends Evented {
...
+     * @example
+     * let popup = new mapboxgl.Popup()
+     * popup.removeClassName('some-class')
+     */
+    removeClassName(className: string) {
+        this._container.classList.remove(className);
+    }
+
+    /**
+     * When only one argument is present toggle the class value.
```
Comment (vakila) [inline src/ui/popup.js]:
```suggestion
     * Add or remove the given class on the popup container, depending on whether the container currently has that class.
```

File: src/ui/popup.js [hunk 8]
```diff
@@ -332,6 +332,47 @@ export default class Popup extends Evented {
...
+     *
+     * @param {string} className non empty string
```
Comment (vakila) [inline src/ui/popup.js]:
```suggestion
     * @param {string} className Non-empty string with CSS class name to add/remove
```

File: test/unit/ui/popup.test.js [hunk 9]
```diff
@@ -495,7 +495,7 @@ test('Popup#remove is idempotent (#2395)', (t) => {
     t.end();
 });
 
-test('Popup adds classes from className option', (t) => {
+test('Popup adds classes from className option, methods for class manipulations works properly', (t) => {
```
Comment (vakila) [inline test/unit/ui/popup.test.js]: Thanks for adding these tests!

Perhaps we should also have tests for edge cases, e.g. if someone passes an empty string (`''`) or a space-separated string (`'multiple class names'`) like they can pass to the `className` option? Would be great to capture the intended behavior for those cases.

File: .gitignore (hunk 1 above)
Comment (Ashot-KR) [reply .gitignore]: `.idea` folder contains project settings for Jetbrains IDEs (Intellij/PhpStrom/WebStorm etc). These IDE pretty popular in webdev, so i thought having it in `.gitingore` could be usefull.
Anyway it's not necessary and could be removed.

File: src/ui/popup.js (hunk 2 above)
Comment (Ashot-KR) [reply src/ui/popup.js]: In previous commit i rewrite foreEach loop to one line function `this._container.classList.add(...name)`(or something like that), but i forgot to check browser compatibility – IE11 does not support multiple arguments, so i revert it to old good `forEach` to keep IE11 working. I noticed it [in this comment](https://github.com/mapbox/mapbox-gl-js/pull/8759#issuecomment-539913416)

Comment (vakila) [reply src/ui/popup.js]: Thanks for the quick reply! Yeah I noticed the revert, I think just the brackets changed from the original (https://github.com/mapbox/mapbox-gl-js/blob/master/src/ui/popup.js#L367-L368) - not really a big deal but no point touching it if we don't need to.

Comment (Ashot-KR) [reply src/ui/popup.js]: Without brackets function returns value, which is pointless for `forEach`. I think it's better use arrow function with brackets here. But as you said it's not a big deal, so i can remove it if you think that we should keep original variant.

File: test/unit/ui/popup.test.js (hunk 9 above)
Comment (Ashot-KR) [reply test/unit/ui/popup.test.js]: My idea for these methods was to keep it closer as possible to native `classList` methods. Native methods throws `DOMException` in these cases(`''` or `'multiple class names'`). And at this point we have 2 options:
1. Just add `t.throws` test for these cases
2. Add checks to my methods to prevent exceptions (e.g. `if (className) { this._container.classList.add(className) }`)

Which one is preferable?

Comment (vakila) [reply test/unit/ui/popup.test.js]: 1 is fine - the idea is to use the tests to capture/document that we've thought about these edge cases & what behavior we expect, it's fine if we expect errors.

File: src/ui/popup.js (hunk 2 above)
Comment (vakila) [reply src/ui/popup.js]: Yes exactly, all this does is change the return value of the function, which isn't used anyway. I don't see this change as necessary/an improvement but you can leave it in if you want to.

**Comment by vakila** [comment]
@Ashot-KR Thanks for getting back to us quickly! Left a few clarifications re the points I raised, but I wouldn't consider those real blockers. Could you resolve the merge conflicts & lint errors so we can get this ready to merge in? Thanks!

File: test/unit/ui/popup.test.js (hunk 9 above)
Comment (Ashot-KR) [reply test/unit/ui/popup.test.js]: tests added

**Comment by Ashot-KR** [comment]
@vakila conflict resolved
also added some changes from conversations above

**Comment by vakila** [comment]
Awesome, thanks a lot @Ashot-KR! Appreciate the extra tests :heart: 

I just noticed it looks like the `addClass` docs updates didn't actually make it into https://github.com/mapbox/mapbox-gl-js/pull/8759/commits/b00e6eb0c7221bb9194477eb443b507300c2d925 - not a huge deal but for consistency with the other methods I just pushed those changes to your fork branch.

So once the checks run I think this is good to go! :rocket: Thanks again for your work on this!

**Comment by Ashot-KR** [comment]
> docs updates didn't actually make it into b00e6eb 

Ooops, something went wrong…
Thanks for fixing that
